### PR TITLE
Add demux operator

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

File: hydroflow/tests/surface_codegen.rs [hunk 1]
```diff
@@ -415,6 +416,75 @@ pub fn test_groupby() {
...
+        Rectangle { width: f64, height: f64 },
+        Square(f64),
+    }
+
+    let mut df = hydroflow_syntax! {
+        demux = recv_iter([
+            Shape::Circle(5.0),
+            Shape::Rectangle { width: 10.0, height: 8.0 },
+            Shape::Square(9.0),
+        ]) -> demux(|shape, tl!(circ, rect)| {
```
Comment (jhellerstein) [inline hydroflow/tests/surface_codegen.rs:432]: tl! is a macro of your creation yes? If we're exposing it to users, I'd prefer a more descriptive name.

File: hydroflow/tests/surface_codegen.rs [hunk 2]
```diff
@@ -415,6 +416,75 @@ pub fn test_groupby() {
...
+        out = merge() -> for_each(|a| println!("area: {}", a));
+
+        demux[circ] -> map(|r| std::f64::consts::PI * r * r) -> out;
+        demux[rect] -> map(|(w, h)| w * h) -> out;
+    };
+    df.run_available();
+}
+
+#[test]
+pub fn test_demux_fizzbuzz_1() {
```
Comment (jhellerstein) [inline hydroflow/tests/surface_codegen.rs:449]: Would be nice to check correct answers for fizzbuzz

File: hydroflow_lang/src/graph/ops/demux.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+use quote::{quote_spanned, ToTokens};
+use syn::spanned::Spanned;
+use syn::{parse_quote, ExprClosure};
+
+// TODO(mingwei): Preprocess rustdoc links in mdbook or in the `operator_docgen` macro.
+
+/// > Arguments: A Rust closure, the first argument is a received item and the
+/// > second argument is a [`tl!` tuple list](https://hydro-project.github.io/hydroflow/doc/hydroflow/macro.tl.html)
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/demux.rs:14]: For me, even tuple_list isn't very helpful ... not sure what that means.

File: hydroflow_lang/src/graph/ops/demux.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+/// Takes the input stream and allows the user to determine what elemnt(s) to
+/// deliver to any number of output streams.
+///
+/// > Note: Downstream operators may need explicit type annotations.
+///
+/// > Note: Import the [`Pusherator`](https://hydro-project.github.io/hydroflow/doc/pusherator/trait.Pusherator.html)
+/// > trait to use the [`.give(...)` method](https://hydro-project.github.io/hydroflow/doc/pusherator/trait.Pusherator.html#tymethod.give).
+///
+/// ```hydroflow
+/// demux = recv_iter(1..=100) -> demux(|v, tl!(fzbz, fizz, buzz, vals)|
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/demux.rs:26]: Confusing to use variable names that are the same as keywords, even if it's legal.

File: hydroflow/tests/surface_codegen.rs [hunk 5]
```diff
@@ -415,6 +416,75 @@ pub fn test_groupby() {
...
+#[test]
+pub fn test_demux_1() {
+    enum Shape {
+        Circle(f64),
+        Rectangle { width: f64, height: f64 },
+        Square(f64),
+    }
+
+    let mut df = hydroflow_syntax! {
+        demux = recv_iter([
```
Comment (jhellerstein) [inline hydroflow/tests/surface_codegen.rs:428]: Confusing to use variable names that are the same as keywords, even if it's legal.

File: hydroflow_lang/src/graph/ops/demux.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                     outputs,
+                     output_ports,
+                     arguments,
+                     is_pull,
+                     ..
+                 }| {
+        assert!(!is_pull);
+        let func = &arguments[0];
+        let func: ExprClosure = parse_quote!(#func);
+        assert_eq!(2, func.inputs.len(), "Closure provided as second argument to `demux` must have a second input argument listing ports, `tl!(port_a, port_b, ...)`.");
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/demux.rs:62]: If you change the name of the macro, be careful to fix it inside strings and comments as well.

File: hydroflow_lang/src/graph/ops/demux.rs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        let input_ports_str = func.inputs[1].to_token_stream().to_string();
+
+        let mut sort_permute: Vec<_> = (0..outputs.len()).collect();
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/demux.rs:65]: confusing variable name, not sure why you use both those words

File: hydroflow_lang/src/graph/ops/demux.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        sort_permute.sort_by_key(|&i| {
+            if let PortIndexValue::Path(expr_path) = output_ports[i] {
+                let port_needle = expr_path.to_token_stream().to_string();
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/demux.rs:68]: another mystical variable name.

File: hydroflow_lang/src/graph/ops/demux.rs (hunk 7 above)
Comment (jhellerstein) [reply hydroflow_lang/src/graph/ops/demux.rs:65]: A little comment on what you're sorting and by what key would be nice.

**Review by jhellerstein — APPROVED** [review]
LGTM. Some of it is a bit opaque, could be easier to read.

File: hydroflow/tests/surface_codegen.rs (hunk 1 above)
Comment (MingweiSamuel) [reply hydroflow/tests/surface_codegen.rs:432]: Will do this separately in #264 since this shows up in a good few other places

**Pull request merged**